pkg/util/status/status.go

```diff
@@ -0,0 +1,300 @@
+package status
+
+import (
+	"reflect"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+)
+
+// updater transparently switches for status updates between /status and the main entrypoint for resource,
+// allowing CRDs to enable or disable the status subresource support anytime.
+type updater struct {
+	lock        sync.Mutex
+	subresource bool
+	cli         kubecli.KubevirtClient
+}
+
+func (u *updater) update(obj runtime.Object) (err error) {
+	if u.getSubresource() {
+		return u.updateWithSubresource(obj)
+	} else {
+		return u.updateWithoutSubresource(obj)
+	}
+}
+
+func (u *updater) patch(obj runtime.Object, pt types.PatchType, data []byte) (err error) {
+	if u.getSubresource() {
+		return u.patchWithSubresource(obj, pt, data)
+	} else {
+		return u.patchWithoutSubresource(obj, pt, data)
+	}
+}
+
+// updateWithoutSubresource will try to update the  status via PUT sent to the main REST endpoint.
+// If status of the returned object did not change, it knows that it should have used the /status subresource
+// and will switch the updater itself over to permanently use the /status subresource.
+func (u *updater) updateWithoutSubresource(obj runtime.Object) (err error) {
+	oldStatus, newStatus, err := u.updateUnstructured(obj)
+	if err != nil {
+		return err
+	}
+	if !reflect.DeepEqual(oldStatus, newStatus) {
+		u.setSubresource(true)
+		return u.updateStatusUnstructured(obj)
+	}
+	return nil
+}
+
+// updateWithSubresource will try to update the  status via PUT sent to the /status subresource.
+// If a 404 error is returned, it will try the main rest entrypoint instead. In case that this
+// call succeeds, it will switch the updater to permanently use the main entrypoint.
+func (u *updater) updateWithSubresource(obj runtime.Object) (err error) {
+	err = u.updateStatusUnstructured(obj)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		oldStatus, newStatus, err := u.updateUnstructured(obj)
+		if errors.IsNotFound(err) {
+			// object does not exist
+			return err
+		}
+		if err == nil && reflect.DeepEqual(oldStatus, newStatus) {
```

Comment:
if somehow the DeepEqual is false here, we need an error. right now, it's possible for the update to succeed, the deep equal to fail, and error is nil. That would make it look like the update succeeded when it did not... I don't know if this situation is even technically possible, but it's worth closing the gap.

Reply:
I think this covers the situation we are interested in. If no error happened but the status is still not in sync, then we initially using the wrong endpoint and we should permanently change to the update method without the subresource. If the error is not nil, or the error is in sync we just return whatever is in err.

Reply:
You were right. Addressed and added a test.